### ISSUE TITLE
Recursive `<expr>` mapping works for builtin mappings but not for custom ones

### ISSUE THREAD

**Issue by echasnovski** [vim/vim#10907]
### Steps to reproduce

I am not particularly sure that it is a bug, but at least this doesn't seem like a consistent behavior.

Steps:
1. Create the following config file (for example, call it 'init_expr-map.vim'):
```vim
function ReturnAc()
  return 'ac'
endfunction

function ReturnIw()
  return 'iw'
endfunction

xnoremap ac 3l
xmap <expr> a ReturnAc()
xmap <expr> A ReturnAc()
xmap <expr> i ReturnIw()
```
2. Start vim with `vim -u init_expr-map.vim init_expr-map.vim`.
3. From Normal mode type `vi`. It selects whole word, as expected because it should be equivalent to `viw`.
4. From Normal mode type `vac` (without big pause between `a` and `c` so that it is recognized from one mapping). It selects equivalent to `v3l`.
5. From Normal mode type `vA`. It selects like `vac` (and `v3l`).
6. From Normal mode type `va`. It doesn't react, however it might be reasonable to expect it to be equivalent to `vac`.

Here is a video demo:

https://user-images.githubusercontent.com/24854248/184499413-01c711a6-f392-4ea0-9c0e-4d215d3d3039.mp4

### Expected behaviour

Typing `va` is identical to (quickly) typing `vac`, i.e. output of expression mapping is executed as if typed with zero delay between keystrokes.

### Version of Vim

9.0.1-135

### Environment

OS: EndeavourOS Linux x86_64 (5.18.16-arch1-1)
Terminal: xfce4-terminal 1.0.4 (Xfce 4.16)
$TERM: xterm-256color
shell: zsh 5.9 (x86_64-pc-linux-gnu)

**Comment by lacygoill** [comment]
When you type `a` in visual mode, Vim waits until the timeout to let you use the `xnoremap ac 3l` mapping.  That is, maybe you intend to type `ac`, and you've only pressed `a`:
```vim
vim9script
'some text'->setline(1)
&showcmd = true
xnoremap ac 3l
xmap a ac
feedkeys('va', 't')
```
Once the timeout occurs, `a` is replaced with `ac` in the typeahead buffer.  But then, `ac` is not replaced with `3l`, because – inside the last used mapping – the RHS starts with the LHS.  That is, in the `xmap a ac` mapping, the RHS `ac` starts with the LHS `a`.  When that happens, Vim does not map the first key again:

   > If the {rhs} starts with {lhs}, the first character is not mapped again

See [`:help recursive_mapping`](https://vimhelp.org/map.txt.html#recursive_mapping).

So, out of the 2 characters `a` and `c` which are written in the typeahead buffer at that point, only `c` can be mapped again.  But you don't have any mapping starting with `c`, so as you said it:

   > It doesn't react

**Comment by echasnovski** [comment]
Thanks for the details! I've encountered this help page, but somehow missed this description, which helps to understand current behavior.

Unfortunately, I still feel that the point stands (maybe because of possibility of ambiguous interpretation). Even if first character is not mapped again, I don't see a reason why first *two* or *three* should not (apart from technical difficulties, of course). That is the case here: `a` is not re-mapped but `ac` might.

I understand that it is a weird edge case, but which proved to be a somewhat important one when trying to make custom mappings for `a` and `i` (basically: wait for user input and fall back to other mappings on improper input).

**Comment by lacygoill** [comment]
> Even if first character is not mapped again, I don't see a reason why first two or three should not (apart from technical difficulties, of course). That is the case here: a is not re-mapped but ac might.

It depends on how you interpret "the first character is not mapped again".
It seems that for you, it just means that the key should not be replaced *alone*.
But for Vim, currently, it means that the key should not be replaced *at all*.

Once a character has been marked as not being re-mappable, it's over; Vim skips it inside the typeahead buffer.

Whether we could or should change this semantics now, I don't know; that's for the devs to decide.

**Comment by zeertzjq** [comment]
You may try returning `"\<Ignore>ac"` instead.

**Comment by echasnovski** [comment]
> You may try returning `"\<Ignore>ac"` instead.

Thanks for this! Although, something weird is going on: it works with `"\<Ignore>ac"`, but hangs with 100% Neovim usage with `"\<Ignore>iw"` (I imagine in infinite recursion because `'maxmapdepth'` doesn't catch it?). I would imagine those cases to be equivalent somehow.

Anyway, I remember trying it in my original use case in Neovim and it didn't work. Now after this suggestion I decided to retry and prepend with `<Ignore>` only user mappings (checked with `maparg()`) and it works! Thanks again!

My initial problem is solved, but I still somehow feel that behavior here should be similar for both user defined and builtin mappings. So, please, feel free to close the issue if that is not the case.

**Comment by zeertzjq** [comment]
Do you mean 100% CPU usage? In this case #10652 may help avoid 100% CPU usage.

**Comment by echasnovski** [comment]
> Do you mean 100% CPU usage? In this case #10652 may help avoid 100% CPU usage.

Yes, 100% usage of single CPU by vim (which can be stopped by `<C-c>`, so not completely hanging). In my case I need this behavior only as fall back, so checking with `maparg()` seems to work at the moment (for Neovim>=0.5, at least).
